### DISCUSSION TITLE
hasAnyTag but not listing of product tags?

### DISCUSSION THREAD

**Discussion by jesper-madsen** [Shopify/function-examples#241] [Feedback: Discounts APIs]
Firstly thanks for making Shopify Functions, and adding support for JS 💚!

After working with Functions for the past week, I found it a limitation that you can check if a product has a specific hasAnyTag().
This is a nice filter/function. But I'm really missing the possibility to read all the tags of a given product in the cart.

So instead of having a true/false if a product has a tag, it would be cool to have all the tags of the product instead to be honest.

**Comment by DuncanUszkay1** [comment]
Hello Jesper 👋🏻 

That would be nice and simple, but for many products that list is too large for us to send. Is there something specific you're trying to accomplish with a complete list of tags that isn't well served by the current API?

**Reply by gil--** [reply]
On my end the problem is that sometimes we want to check for multiple tags and there's no way to know which tag it has, right? Would just be true/false for the list.

**Reply by DuncanUszkay1** [reply]
It is possible to call `hasAnyTag` once per tag you're looking for using GraphQL aliases, but that's admittedly expensive in terms of the input query budget. It's an issue we're aware of 👀

**Reply by gil--** [reply]
Yea in my use case i've created a generalized app for making customizations and so it supports a lot of different inputs but that means it's using the entire 30/30 input query cost with just a single hasAnyTag.

**Reply by jesper-madsen** [reply]
The reason I'm asking for a full list of product tags is because we have multiple clients that use Shopify Scripts. Metafields was not supported in Scripts so we relied on a tag structure like `tier_{qty}_{percent_discount}` eg. `tier_4_20` to pass discount information to the script.

But as you point out that it is quite compute intensive, we will just migrate to using a single metafield instead of relying on tags to pass on information.

**Reply by bsmithrainbowshops** [reply]
Our store has the same issue as @daddy88  for the same reason, we put our promos in a product tag to be picked up by Shopify Scripts for line-item discounting. Example, every product with a promo has something along the lines of: "promotext: X_for_Y"

I suspect we'll also have to migrate to a metafield on each product and check against those. Any suggestions as to whether this is the most efficient way to do this in Functions...?

**Reply by DuncanUszkay1** [reply]
>I suspect we'll also have to migrate to a metafield on each product and check against those. Any suggestions as to whether this is the most efficient way to do this in Functions...?

Yes, parsing those tags and writing the results into metafields would be the way to go here 👍🏻

**Reply by sethbeer** [reply]
We have a similar issue as many of our clients use tags to identify loyalty tiers and pricing. Our clients can make a new tier for their loyalty programs just by creating a new tag in a predefined format. Moving forward we will be required to create a new extension and deploy it to allow the new tag to be used or we will need to switch all functionality over to metafields. Most of our clients stipulate that they should be able to create new loyalty tiers without needing a deployment. We also use tags on the front end to show different views, all of this would also need shifting to use metafields which will be a huge undertaking.

**Reply by bsmithrainbowshops** [reply]
@sethbeer I think everyone on Shopify is going to have to undergo that "huge undertaking", migrating all tags that are specific to checkout functionality over to metafields. We've already started to do this in our product deploys, Matrixify can help do it for current products, but what you're saying is not a joke: it is a huge undertaking. 
I'd like to see "recommended patterns for migration" or "migration cookbook" that mentions: "stop using product tags, start using product metafields".

**Reply by jasonplumb** [reply]
We also have the same on our store e.g. `discount:100` where we lookup the key `discount` and parse the value `100`. I can see how the metafield approach is the only way when dealing with shopify functions but I'm unclear why we can't just have the following graphql query:

```graphql
product {
 tags
}
```

In liquid you can retrieve all the product tags and in the worse case scenario the existing logic for `hasAnyTag` would be `O(n)` so why not have a response with all the tags sent?

**Comment by danbisso** [comment]
For those of us migrating from Scripts that require checking a _list_ of product tags without eating up our query cost allowance, there's another potential solution that was [introduced in a recent API version](https://shopify.dev/docs/api/functions/reference/product-discounts/graphql/common-objects/product?api%5Bversion%5D=2023-07). Instead of using `hasAnyTag` to check a single tag, you can use the newer `hasTags` and provide a list of tags to it via your Function's **input query** params. E.g:
```
query Input($tags: [String!]!) {
  cart {
    lines {
      merchandise {
        ...on ProductVariant {
          product {
            hasTags(tags: $tags) {
              hasTag
              tag
            }
          }
        }
      }
    }
  }
}
```

Shopify will perform the comparison for you and return [a data structure](https://shopify.dev/docs/api/functions/reference/product-discounts/graphql/common-objects/hastagresponse?api%5Bversion%5D=2023-07) containing each tag you provided and whether or not that tag is assigned to the product. E.g:

```
"hasTags":  [
  {
    "hasTag": true,
    "tag": "a-product-tag"
  },
  {
    "hasTag": false,
    "tag": "another-tag"
  }
]
```

It is a bit cumbersome, but there are many benefits to this approach: you don't have to hard-code specific tags in your input query, its query cost is just '3' for checking _multiple tags_, and you don't have to migrate tags to metafields.

**Comment by nickwesselman** [comment]
As mentioned above, the `hasTags` field was introduced to cover this use case.
